### ISSUE TITLE
[plugin|tiling] Window resize animation

### ISSUE THREAD

**Issue by dominiklohmann** [koekeishiya/chunkwm#185]
Just a random thought.

NSWindow has functions for smooth window resizes:
- [`setFrame:display:animate:`](https://developer.apple.com/documentation/appkit/nswindow/1419519-setframe?language=objc)
- [`animationResizeTime:`](https://developer.apple.com/documentation/appkit/nswindow/1419655-animationresizetime?language=objc)

This in combination with code injection (e.g. when using chwm-sa) could be used to make windows resize smoothly.

**Comment by koekeishiya** [comment]
There is no way as far as I know to go from a cgswindowid to NSWindow reference

**Comment by dominiklohmann** [comment]
If I understood the few Stackoverflow discussions available on that topic, the following might work (if the window in question is of type `NSWindow` ). Apparently window number and `CGWindowID` are essentially the same for `NSWindow`.

```objc
CGWindowID id = ...
NSWindow *window = [[NSApplication sharedApplication] windowWithWindowNumber: (NSInteger)id];
```

**Comment by choco** [comment]
Yep, I actually explained how to do it to a user on the irc channel, and he reported that it was working allright.
The main issue is that some applications enforce `Library Validation` which basically checks if the loaded library is either signed by Apple, or by the same developer of the applications. This unfortunately makes Scripting Addition fail loading in these apps, and unfortunately it is becoming more widely used :( For example Safari implements this. I'm actually writing a kernel extension to help with this, but I'm not sure it would the right solution for everyone....

**Comment by koekeishiya** [comment]
@dominiklohmann 

I'm pretty sure that would require us to inject our scripting addition into specific applications like SIMBL plugins does. It cannot be done solely through *chwm-sa* and the Dock.

I have not spent much time looking into how scripting additions actually work and what kind of limits an application developer can impose, so @choco is probably the better person to answer these types of questions.

**Comment by dominiklohmann** [comment]
Damn. All I want is resize and reposition to happen in the same frame. Seems like not even the deprecated `NSDisableScreenUpdates()` / `NSEnableScreenUpdates()` or the replacement in form of `CATransition` can do it for another processes windows.

**Comment by koekeishiya** [comment]
@dominiklohmann 

The Dock has a rerefence to some type called "FSWMWindow" and it has a  few methods that seem interesting

initWithWindow looks like it takes one of our CGWindowIDs, so that's good.
"forceSetFrame" makes me think it can actually bypass minimum/maximum size constraints (and does move + resize) ??

```
 7540 - (id)initWithWindow:(unsigned int)v1;
 7541 - (void)forceSetFrame:(struct CGRect)v1;
```
